### ISSUE TITLE
Understanding SizeKB attribute of DRAM memory

### ISSUE THREAD

**Issue by suyashbakshi** [NVlabs/timeloop#156]
Hello, I'm trying to understand the sizeKB attribute of DRAM class and am observing weird behavior. I've attached the arch.yaml, prob_3d.yaml (3D conv problem) and a sample mapping.yaml.
When attempting to evaluate the sample mapping attached using timeloop-model:
- when I set sizeKB for DRAM = 102400, the error I get is
```ERROR: couldn't map level MainMemory: mapped tile size 669141632 exceeds buffer capacity 104857600```

- when I set sizeKB = 409600, the error is:
```ERROR: couldn't map level MainMemory: mapped tile size 669141632 exceeds buffer capacity 419430400```

- when setting sizeKB = 1638400
```ERROR: couldn't map level MainMemory: mapped tile size 669141632 exceeds buffer capacity 67108864```

My questions are:
1. The total number of data elements in the layer are 669141632, which for 8-bit elements is ~638MegaBytes. So this should fit in the MainMemory when sizeKB is set to 1638400 (which should be 1600MegaBytes).
2. If you notice the errors above, when setting sizeKB = 1638400, the error message says the buffer capacity is 67108864, which is less than the buffer capacity reported in error message when sizeKB is set to 409600 (buffer capacity reported in error = 419430400). Why is the buffer capacity smaller for the larger sizeKB?

All of this is assuming the "sizeKB" stands for "size in kilo Bytes". 

I'd appreciate if you could explain this behavior to me, and correct me if I'm understanding the sizeKB attribute incorrectly.

[arch.txt](https://github.com/NVlabs/timeloop/files/9550923/arch.txt)
[prob_3d.txt](https://github.com/NVlabs/timeloop/files/9550930/prob_3d.txt)
[mapping.txt](https://github.com/NVlabs/timeloop/files/9550935/mapping.txt)

**Comment by angshuman-parashar** [comment]
Looks like a bug. Reported "capacity" is correct for your first two invocations, but went down in the last invocation. There's clearly something wrong. Could be an integer overflow. My suspicion is it's this line: https://github.com/NVlabs/timeloop/blob/c71b11f57dd2189ca5ea04f36343fc4308b011c0/src/model/buffer.cpp#L324

Could you try changing the uint32 to uint64 here and see if it fixes the problem: https://github.com/NVlabs/timeloop/blob/c71b11f57dd2189ca5ea04f36343fc4308b011c0/src/model/buffer.cpp#L308

Thanks!

**Comment by suyashbakshi** [comment]
Hello, that didn't work out. I changed the variable "size" to uint64, but that generated following errors:
```
src/model/buffer.cpp: In static member function ‘static model::BufferLevel::Specs model::BufferLevel::ParseSpecs(config::CompoundConfigNode, uint64_t, bool)’:
src/model/buffer.cpp:288:25: error: no matching function for call to ‘lookupValue(const char [8], uint64_t&)’
  288 |   if (buffer.lookupValue("entries", size) )
      |       ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
In file included from include/workload/shape-models/problem-shape.hpp:34,
                 from include/mapping/loop.hpp:36,
                 from include/loop-analysis/tiling.hpp:32,
                 from include/model/level.hpp:31,
                 from include/model/buffer.hpp:34,
                 from src/model/buffer.cpp:36:
include/compound-config/compound-config.hpp:58:8: note: candidate: ‘bool config::CompoundConfigNode::lookupValue(const char*, bool&) const’ (near match)
   58 |   bool lookupValue(const char *name, bool &value) const;
      |        ^~~~~~~~~~~
include/compound-config/compound-config.hpp:58:8: note:   conversion of argument 2 would be ill-formed:
``` 

In an attempt to resolve them, I overloaded the "lookup_value" function in "include/compund-config/compound-config.hpp" for "uint64_t" but that resulted in following error when building:
```
/usr/bin/ld: build/libtimeloop-model.so: undefined reference to `config::CompoundConfigNode::lookupValue(char const*, unsigned long&) const'
collect2: error: ld returned 1 exit status
scons: *** [build/timeloop-design-space] Error 1
scons: building terminated because of errors.
```

**Comment by angshuman-parashar** [comment]
Try this — leave CompoundConfig unchanged. Capture the value (ie lookupvalue) into a uint32 but immediately upcast it to a uint64 so that all the math happens in 64 bit.

**Comment by suyashbakshi** [comment]
Thank you. That worked. I modified the statement:
https://github.com/NVlabs/timeloop/blob/c71b11f57dd2189ca5ea04f36343fc4308b011c0/src/model/buffer.cpp#L324

to
```
specs.size = (uint64_t)size * 1024 * 8 / specs.word_bits.Get();
```
which resulted in expected behavior for me.

**Comment by angshuman-parashar** [comment]
Fixed in b7476e77df519e27e8c8b0b6bbf29c83f5112925.
